dali/operators/reader/loader/tar_utils.cc

```diff
@@ -0,0 +1,189 @@
+// limitations under the License.
+
+#include "dali/operators/reader/loader/tar_utils.h"
+#include <libtar.h>
+#include <algorithm>
+#include <cstdarg>
+#include <cstdlib>
+#include <list>
+#include <string>
+#include <tuple>
+#include <utility>
+#include "dali/core/error_handling.h"
+
+namespace dali {
+namespace detail {
+
+namespace {
+
+constexpr uint64_t kBlockSize = 512;
+constexpr uint64_t kEmptyEofBlocks = 2;
+constexpr uint64_t kTarArchiveBufferInitSize = 1;
+
+std::mutex instances_mutex;
+std::list<std::vector<TarArchive*>> instances_registry = {
+    std::vector<TarArchive*>(kTarArchiveBufferInitSize)};
+TarArchive** instances = instances_registry.back().data();
+
+}  // namespace
+
+int Register(TarArchive* archive) {
+  std::lock_guard<std::mutex> instances_lock(instances_mutex);
+  for (auto& instances_entry : instances_registry.back()) {
+    if (instances_entry != nullptr) {
+      continue;
+    }
+    instances_entry = archive;
+    return &instances_entry - instances;
+  }
+  std::vector<TarArchive*>& old = instances_registry.back();
+  instances_registry.emplace_back();
+  std::vector<TarArchive*>& curr = instances_registry.back();
+  curr.reserve(old.size() * 2);
+  curr = old;
+  curr.push_back(archive);
+  curr.resize(old.size() * 2, nullptr);
+  instances = curr.data();
+  return old.size();
+}
+
+inline void Unregister(int instance_handle) {
+  instances[instance_handle] = nullptr;
+}
+
+TarArchive::TarArchive(std::unique_ptr<FileStream> stream)
+    : stream(std::move(stream)), eof(false), instance_handle(Register(this)), archiveoffset(0) {
+  this->stream->Seek(0);
+  ParseHeader();
+}
+
+TarArchive::TarArchive(TarArchive&& other) {
+  *this = static_cast<TarArchive&&>(other);
+}
+
+TarArchive::~TarArchive() {
+  Unregister(instance_handle);
+}
+
+TarArchive& TarArchive::operator=(TarArchive&& other) {
+  stream = std::move(other.stream);
+  eof = other.eof;
+  filename = other.filename;
+  filesize = other.filesize;
+  readoffset = other.readoffset;
+  archiveoffset = other.archiveoffset;
+
+  std::lock_guard<std::mutex> instances_lock(instances_mutex);
+  instances[instance_handle] = nullptr;
+  instance_handle = other.instance_handle;
+  return *this;
+}
+
+uint64_t RoundToBlockSize(uint64_t count) {
+  return ((count + kBlockSize - 1) / kBlockSize) * kBlockSize;
+}
+
+bool TarArchive::NextFile() {
+  if (eof) {
+    return false;
+  }
+  Skip(RoundToBlockSize(filesize) - readoffset);
+  eof = ParseHeader();
+  return !eof;
+}
+
+bool TarArchive::IsAtFile() const {
+  return !eof;
+}
+
+std::string TarArchive::GetFileName() const {
+  return filename;
+}
+
+uint64_t TarArchive::GetFileSize() const {
+  return filesize;
+}
+
+std::shared_ptr<uint8_t> TarArchive::ReadFile() {
+  archiveoffset -= readoffset;
+  readoffset = 0;
+  stream->Seek(archiveoffset);
+  auto buffer = std::shared_ptr<uint8_t>(new uint8_t[filesize], [](uint8_t* ptr) { delete[] ptr; });
+  Read(buffer.get(), filesize);
```

Comment:
That's definitely not what we want.
```suggestion
  buffer = stream->Get(filesize);
  stream->Seek(/* whatever the offset to the next header is */);
  return buffer;
```

Reply:
but for some readers it's not implemented (like for StdFileStream)